bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableIO.java

```diff
@@ -0,0 +1,1039 @@
+      this.session = session;
+    }
+
+    @VisibleForTesting
+    protected void setScanner(ResultScanner<FlatRow> scanner) {
+      this.scanner = scanner;
+    }
+
+    @VisibleForTesting
+    public ByteKeyRangeTracker getRangeTracker() {
+      return rangeTracker;
+    }
+
+    /**
+     * Closes the {@link ResultScanner}, {@link Table}, and {@link Connection}.
+     */
+    @Override
+    public void close() throws IOException {
+      scanner.close();
+      long totalOps = getRowsReadCount();
+      long elapsedTimeMs = System.currentTimeMillis() - workStart;
+      long operationsPerSecond = elapsedTimeMs == 0 ? 0 : (totalOps * 1000 / elapsedTimeMs);
+      READER_LOG.info(
+          "{} Complete: {} operations in {} ms. That's {} operations/sec",
+          this,
+          totalOps,
+          elapsedTimeMs,
+          operationsPerSecond);
+    }
+
+    @VisibleForTesting
+    long getRowsReadCount() {
+      return rowsRead.get();
+    }
+
+    @Override
+    public final Result getCurrent() throws NoSuchElementException {
+      return current;
+    }
+
+    @Override
+    public final synchronized BoundedSource<Result> getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public String toString() {
+      return String.format("Reader for: ['%s' - '%s']",
+        Bytes.toStringBinary(rangeTracker.getStartPosition().getBytes()),
+        Bytes.toStringBinary(rangeTracker.getStopPosition().getBytes()));
+    }
+  }
+
+  ///////////////////// Write Class /////////////////////////////////
+
+
+  private static class MutationStatsExporter {
+    protected static final Logger STATS_LOG = LoggerFactory.getLogger(AbstractSource.class);
+    private static MutationStatsExporter mutationStatsExporter;
+
+    static synchronized void initializeMutationStatsExporter() {
+      if (mutationStatsExporter == null) {
+        mutationStatsExporter = new MutationStatsExporter();
+        mutationStatsExporter.startExport();
+      }
+    }
+
+    protected void startExport() {
+      Runnable r = new Runnable() {
+        @Override
+        public void run() {
+          try {
+            BufferedMutatorDoFn.cumulativeThrottlingSeconds.set(TimeUnit.NANOSECONDS
+                .toSeconds(ResourceLimiterStats.getInstance().getCumulativeThrottlingTimeNanos()));
+          } catch (Exception e) {
+            STATS_LOG.warn("Something bad happened in export stats", e);
+          }
+        }
+      };
+      BigtableSessionSharedThreadPools.getInstance().getRetryExecutor().scheduleAtFixedRate(r, 5, 5,
+        TimeUnit.MILLISECONDS);
+    }
+  }
+
+  /**
+   * This is a DoFn that relies on {@link BufferedMutator} as the implementation to write data to
+   * Cloud Bigtable. The main function of this class is to manage Aggregators relating to mutations.
+   * @param <InputType>
+   */
+  private static abstract class BufferedMutatorDoFn<InputType> extends AbstractCloudBigtableTableDoFn<InputType, Void> {
```

Comment:
Whats the distinction between AbstractCloudBigtableTableDoFn & BufferedMutatorFn? What other subclasses will there be?